### PR TITLE
O11y rerun

### PR DESCRIPTION
**Pull request by darpanLalwani**

### CODE REVIEW

File: setup-env/src/actionInput/index.js [hunk 1]
```diff
@@ -30,6 +33,11 @@ class ActionInput {
       // non-compulsory fields
       this.buildName = core.getInput(INPUT.BUILD_NAME);
       this.projectName = core.getInput(INPUT.PROJECT_NAME);
+      this.githubToken = core.getInput(INPUT.GITHUB_TOKEN);
+      this.githubApp = core.getInput(INPUT.GITHUB_APP);
+      this.rerunAttempt = process?.env?.GITHUB_RUN_ATTEMPT;
```
Comment (darpanLalwani) [inline setup-env/src/actionInput/index.js:38]: check if available in github context

File: setup-env/src/actionInput/index.js [hunk 2]
```diff
@@ -42,12 +50,14 @@ class ActionInput {
...
     this.projectName = InputValidator.validateProjectName(this.projectName);
+    this.githubToken = InputValidator.validateGithubToken(this.githubToken);
+    this.githubApp = InputValidator.validateGithubAppName(this.githubApp);
   }
 
   /**
    * Sets env variables to be used in the test script for BrowserStack
    */
-  setEnvVariables() {
+  async setEnvVariables() {
```
Comment (darpanLalwani) [inline setup-env/src/actionInput/index.js:60]: Check backward compatibility

File: setup-env/src/actionInput/index.js [hunk 3]
```diff
@@ -64,8 +74,71 @@ class ActionInput {
...
 
+    if (await this.checkIfBStackReRun()) {
+      await this.setBStackRerunEnvVars();
+    }
     core.endGroup();
   }
+
+  async checkIfBStackReRun() {
+    // Using !! ensures that the function returns true or false, regardless of the input values.
+    core.info(`The variables set are: rerunAttempt - ${this.rerunAttempt}, runId - ${this.runId}, repository - ${this.repository}, githubToken - ${this.githubToken}`);
```
Comment (darpanLalwani) [inline setup-env/src/actionInput/index.js:85]: remove this line

File: setup-env/src/actionInput/index.js [hunk 4]
```diff
@@ -64,8 +74,71 @@ class ActionInput {
...
+    try {
+      // Check if the run was triggered by the BrowserStack rerun bot
+      core.info('The re-run was triggered by the GitHub App from BrowserStack.');
+
+      const browserStackApiUrl = BROWSERSTACK_TEMPLATE.DETAILS_API_URL.replace('{runId}', this.runId);
+
+      // Call BrowserStack API to get the tests to rerun
+      const bsApiResponse = await axios.get(browserStackApiUrl, {
+        auth: {
+          username: this.username.replace("-GitHubAction", ""),
```
Comment (darpanLalwani) [inline setup-env/src/actionInput/index.js:124]: Get context for this

File: setup-env/test/actionInput/index.test.js [hunk 5]
```diff
@@ -19,6 +19,7 @@ describe('Action Input operations for fetching all inputs, triggering validation
       sinon.stub(InputValidator, 'updateUsername').returns('validatedUsername');
       sinon.stub(InputValidator, 'validateBuildName').returns('validatedBuildName');
       sinon.stub(InputValidator, 'validateProjectName').returns('validatedProjectName');
+      process.env.GITHUB_REPOSITORY = 'browserstack/github-actions';
```
Comment (darpanLalwani) [inline setup-env/test/actionInput/index.test.js:22]: add test cases for newly added methods

File: setup-env/src/actionInput/index.js (hunk 3 above)
Comment (darpanLalwani) [reply setup-env/src/actionInput/index.js:85]: done

File: setup-env/src/actionInput/index.js (hunk 4 above)
Comment (darpanLalwani) [reply setup-env/src/actionInput/index.js:124]: This is for internal instrumentation. More details added https://browserstack.slack.com/archives/CPHUK7ZA5/p1727083890517749?thread_ts=1726650791.006209&cid=CPHUK7ZA5

File: setup-env/src/actionInput/index.js (hunk 1 above)
Comment (darpanLalwani) [reply setup-env/src/actionInput/index.js:38]: Done, github context does not have rerunAttempt, runId but process does, using that for consistency now. It also removes the import for github

File: setup-env/src/actionInput/index.js (hunk 2 above)
Comment (darpanLalwani) [reply setup-env/src/actionInput/index.js:60]: The calling methods are already async, added await for these method calls. The functionality seems to be working fine anyways.

File: setup-env/config/constants.js [hunk 6]
```diff
@@ -12,4 +14,8 @@ module.exports = {
     BROWSERSTACK_BUILD_NAME: 'BROWSERSTACK_BUILD_NAME',
     BROWSERSTACK_PROJECT_NAME: 'BROWSERSTACK_PROJECT_NAME',
   },
+
+  BROWSERSTACK_TEMPLATE: {
```
Comment (agrawalsaurabhs) [inline setup-env/config/constants.js:18]: Let's call this BROWSERSTACK_INTEGRATIONS

File: setup-env/action.yml [hunk 7]
```diff
@@ -13,6 +13,14 @@ inputs:
   project-name:
     description: 'Project name for the tests'
     required: false
+  github-token:
```
Comment (agrawalsaurabhs) [inline setup-env/action.yml:16]: https://docs.github.com/en/actions/security-for-github-actions/security-guides/automatic-token-authentication#about-the-github_token-secret:~:text=The%20token%20is%20also%20available%20in%20the%20github.token%20context.%20For%20more%20information%2C%20see%20%22Accessing%20contextual%20information%20about%20workflow%20runs.%22

github_token is available as part of github context as per above docs. Can we check once?

**Review by agrawalsaurabhs — CHANGES_REQUESTED** [review]

File: setup-env/src/actionInput/index.js (hunk 1 above)
Comment (agrawalsaurabhs) [reply setup-env/src/actionInput/index.js:38]: https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/accessing-contextual-information-about-workflow-runs#example-printing-context-information-to-the-log:~:text=the%20workflow%20run.-,github.run_attempt,-string

lets check above once.

File: setup-env/config/constants.js (hunk 6 above)
Comment (darpanLalwani) [reply setup-env/config/constants.js:18]: done

Comment (agrawalsaurabhs) [reply setup-env/config/constants.js:18]: let's take handshake with Shirish on username/accesskey and not 2 step authentication.

File: setup-env/test/actionInput/index.test.js (hunk 5 above)
Comment (darpanLalwani) [reply setup-env/test/actionInput/index.test.js:30]: added

File: setup-env/action.yml (hunk 7 above)
Comment (darpanLalwani) [reply setup-env/action.yml:16]: yes, it is available, removed this

File: setup-env/src/actionInput/index.js [hunk 8]
```diff
@@ -64,8 +71,72 @@ class ActionInput {
...
+          password: this.accessKey,
+        },
+        headers: {
+          'Content-Type': 'application/json',
+          'x-rs-auth': true,
```
Comment (darpanLalwani) [inline setup-env/src/actionInput/index.js:127]: remove x-rs-auth

File: setup-env/src/actionInput/index.js (hunk 1 above)
Comment (agrawalsaurabhs) [reply setup-env/src/actionInput/index.js:36]: Using process.env as context is available at action level which doesn't have these details. Accessing from Process.env

File: setup-env/test/actionInput/inputValidator.test.js [hunk 9]
```diff
@@ -152,6 +152,29 @@ describe('InputValidator class to validate individual fields of the action input
           expect(InputValidator._getBuildInfo()).to.eq(expectedValue);
         });
       });
+
```
Comment (agrawalsaurabhs) [inline setup-env/test/actionInput/inputValidator.test.js:155]: Please share prev test coverage vs current for comparision.

File: setup-env/config/constants.js (hunk 6 above)
Comment (darpanLalwani) [reply setup-env/config/constants.js]: Done, shared slack thread with you

File: setup-env/test/actionInput/inputValidator.test.js (hunk 9 above)
Comment (darpanLalwani) [reply setup-env/test/actionInput/inputValidator.test.js:155]: Added on slack thread

Comment (darpanLalwani) [reply setup-env/test/actionInput/inputValidator.test.js:155]: Shared on chat

File: setup-env/config/constants.js [hunk 10]
```diff
@@ -12,4 +14,8 @@ module.exports = {
     BROWSERSTACK_BUILD_NAME: 'BROWSERSTACK_BUILD_NAME',
     BROWSERSTACK_PROJECT_NAME: 'BROWSERSTACK_PROJECT_NAME',
   },
+
+  BROWSERSTACK_INTEGRATIONS: {
+    DETAILS_API_URL: 'https://integrate.browserstack.com/api/ci-tools/v1/builds/{runId}/rebuild/details?tool=github-actions&as_bot=true',
```
Comment (agrawalsaurabhs) [inline setup-env/config/constants.js:19]: tool = github?

**Review by agrawalsaurabhs — DISMISSED** [review]

**Review by pranavj1001 — APPROVED** [review]

**Review by pranavj1001 — APPROVED** [review]

**Pull request merged**